errlog rewrite w/ color removal

mdavidsaver opened the pull request:
Following up the digression on https://github.com/epics-base/epics-base/pull/109#issuecomment-781816959.  There are two parts to this PI.  A re-write of errlog around a double buffer, and teaching errlog to strip out ANSI terminal CSI escape sequences before passing along to log handlers.

The "WIP: add some color" commit is intended as a demo, and should not be merged.  Starting softIoc on Linux should show a blue `iocInit` when `$TERM` is set, and not when it is cleared.  In either case log clients should see plain text without escapes.

The buffering of log entries is conceptually similar, with the additional of a second buffer to allow errlogThread to print without holding a mutex.  This might be a benefit by itself wrt. PI and latency.  I was able to rewrite eg. `errlogPrintf()` to simply delegate to `errlogVprintf()` to avoid duplication, although this savings in LoC is offset by adding `errlogStripANSI()` (which seems in need of improvement).

The only change I needed to make to the errlog unittest was to allow for an additional message in one case.  I think this is due to the added capacity of double buffering.

I suspect that the logic in `isATTY()` which is testing `$TERM` may need to be excluded on RTEMS/vxWorks, but I haven't checked.  As it stands, escapes will be striped unless `$TERM` is set to a non-empty string.

The main omission I can see is that `printf()`, `epicsPrintf()`, and friends aren't covered by this PR.  I couldn't see a straightforward and portable way to do this.

I also haven't investigated the the support for ANSI escapes in Windows 10.

mdavidsaver committed:
Com: rewrite errlog

Switch to double buffering to allow errlogThread
to unlock while printing.

mdavidsaver committed:
errlog strip ANSI escapes

Always strip for handlers, and conditionally
if stderr is not a TTY, or $TERM unset/empty.

mdavidsaver commented:
![image](https://user-images.githubusercontent.com/4429215/108663309-1ebb5d00-7485-11eb-9f8b-5a81e066c9fc.png)

mdavidsaver committed:
errlog: eltc() re-add flush

Removal upsets dbCaLinkTest on RTEMS, which must not be
synchronizing correctly.  Re-add until this can be corrected.

mdavidsaver commented:
Weird.  Removing the `errlogFlush()` from `eltc()` causes `dbCaLinkTest()` to hang on RTEMS.  Clearly there is a synchronization problem with this test which is being hidden by the flush.  I can't recall ever seeing sporadic failures on Linux or Windows though.  I've re-added the flush until this can be investigated further.

`eltc()` no longer needs to flush since the `toConsole` flag is passed through the queue with each message.

FreddieAkeroyd commented:
Windows 10 does support ANSI escapes in some terminals, @dirk-zimoch has added coloured output to Stream Device and some recent changes to make this on/off configurable https://github.com/paulscherrerinstitute/StreamDevice/pull/64 will point you at the lines needed to check for and enable it on Windows 10

dirk-zimoch commented:
How to enable ANSI colors in Windows: https://github.com/paulscherrerinstitute/StreamDevice/blob/ea8873becd656c33aa230731d3890d57cd28ace1/src/StreamError.cc#L52

mdavidsaver committed:
errlog add ANSI escape macros

mdavidsaver committed:
colorize errors and warnings

Use ERL_ERROR and ERL_WARNING

git grep -li 'errlogPrintf.*[" ]error' | xargs sed -i -E -e 's|(errlogPrintf.*[" ])(error)|\1" ERL_ERROR "|g'

git grep -li 'errlogPrintf.*[" ]warn' | xargs sed -i -E -e 's|(errlogPrintf.*[" ])(warn[a-zA-Z]*)|\1" ERL_WARNING "|g'

mdavidsaver committed:
errlog: try to enable WIN10 terminal escape processing

mdavidsaver commented:
Updated.  I've made an untested attempt at enabling windows 10 terminal escapes.  I've also done a search/replace to colorize "error" and "warning" in many internal error messages.  I picked red and magenta since these have good contrast on both dark and light backgrounds.  (recent gcc does the same)

This update introduces a bunch of `ANSI_*` macros to `errlog.h` for specific colors.  I'm not certain if this is worthwhile, so these could easily be omitted.  The internal usage is `ERL_ERROR` and `ERL_WARNING` for these two specific usage contexts, and so could accommodate other formatting (__bold__, _italic_, etc.).

```c
// equivalent
errlogPrintf(ERL_ERROR ": something is amiss\n");
errlogPrintf(ANSI_RED("ERROR") ": something is amiss\n");
errlogPrintf(ANSI_ESC_RED "ERROR" ANSI_ESC_RESET ": something is amiss\n");
errlogPrintf("\033[31;1mERROR\033[0m: something is amiss\n");
```

![dark](https://user-images.githubusercontent.com/4429215/113517426-c7ed7c80-9534-11eb-8b5c-bf17da4860ae.png)
![light](https://user-images.githubusercontent.com/4429215/113517433-cb810380-9534-11eb-9c5b-069fce09458b.png)

mdavidsaver commented:
As discussed during the last Core meeting, I'm planning to merge this change now that `R7.0.6.1` has been tagged.

This also changes the `ring buffer full` message to explicitly say "error".

https://github.com/epics-base/epics-base/blob/78b41f5bbf63381afbbb63ed69fe6a4d1481d839/modules/database/src/ioc/db/callback.c#L88-L90

FreddieAkeroyd commented:
I ran on my windows desktop, seemed to work as planned (coloured output on screen, not coloured to file/procserv)

mdavidsaver commented:
Merged as of b9899213d4e61de8cc28e0006bf5196e605f676e

dirk-zimoch commented on modules/libcom/test/epicsErrlogTest.c:
```diff
@@ -370,7 +403,8 @@ MAIN(epicsErrlogTest)
     epicsEventMustWait(pvt.done);
-    testEqInt(pvt.count, N+1);
+    /* Expect N+1 messages +- 1 depending on impl */
+    testOk(pvt.count >= N && pvt.count<=N+2, "Logged %u messages, expected %zu", pvt.count, N+1);
```
Another %z. I suggest to use unsigned long for counters.

dirk-zimoch commented on modules/libcom/src/error/errlog.c:
```diff
@@ -524,172 +597,97 @@ int errlogInit2(int bufsize, int maxMsgSize)
+
+                if((base[0]&ERL_STATE_MASK) != ERL_STATE_READY || mlen>=pvt.bufSize - pos) {
+                    fprintf(stderr, "Logic Error: errlog buffer corruption. %02x, %zu\n",
+                            (unsigned)base[0], mlen);
```
VxWorks does not support %z, not even in version 6.9.
I suggest to use unsigned int for mlen, pos, bufsize, maxMsgSize. On all supported architectures, unsigned int is large enough for MAX_MESSAGE_SIZE and BUFFER_SIZE.

dirk-zimoch commented on modules/libcom/src/error/errlog.c:
```diff
@@ -524,172 +597,97 @@ int errlogInit2(int bufsize, int maxMsgSize)
-        if (epicsEventTryWait(pvtData.flush) != epicsEventWaitOK)
-            continue;
+            if(nLost && console)
+                fprintf(console, "errlog: lost %zu messages\n", nLost);
```
%z again. Is suggest to use unsigned long for nLost.
